doc/src/guides/solving/solve-polynomial.md

```diff
@@ -0,0 +1,145 @@
+>>> solve(a*x**2 + b*x + c, x, dict=True)
+[{x: (-b - sqrt(-4*a*c + b**2))/(2*a)}, {x: (-b + sqrt(-4*a*c + b**2))/(2*a)}]
+```
+
+This example reproduces the [quadratic
+formula](https://en.wikipedia.org/wiki/Quadratic_formula).
+
+## Guidance
+
+### *Guidance 1*
+
+*Guidance 1 content*
+
+### *Guidance 2*
+
+*Guidance 2 content*
+
+
+## Solve (Find the Roots of) a Polynomial Algebraically
+
+You can solve a polynomial algebraically in several ways. The one to use depends
+on whether you
+- want an algebraic or numeric answer
+- want the multiplicity of each root (how many times each root is a solution).
+  In the `expression` below representing $(x+2)^2(x-3)$, the root -2 has a
+  multiplicity of two because $x+2$ is squared, whereas 3 has a multiplicity of
+  one because $x-3$ has no exponent. Similarly, for the `symbolic` expression,
+  the root $-a$ has a multiplicity of two and the root $b$ has a multiplicity of
+  one.
+
+```py
+>>> from sympy import solve, roots, real_roots, factor, nroots, RootOf
+>>> from sympy import Poly
+>>> from sympy.abc import x, a, b
+>>> expression = (x+2)**2 * (x-3)
+>>> symbolic = (x+a)**2 * (x-b)
+```
+
+### Algebraic Solution Without Root Multiplicities
+
+You can use SymPy's standard {func}`~.solve` function, though it will not return
+the multiplicity of roots:
+
+```py
+>>> solve(expression, x, dict=True)
+[{x: -2}, {x: 3}]
+>>> solve(symbolic, x, dict=True)
+[{x: -a}, {x: b}]
+```
+
+Refer to [](solve-equation-algebraically.md) for more about using
+{func}`~.solve`.
+
+### Algebraic Solution With Root Multiplicities
+
+#### {func}`~.roots`
+
+{func}`~.roots` is the most rigorous function because it can give explicit
+expressions for the roots of polynomials that have symbolic coefficients (that
+is, if there are symbols in the coefficients) if {func}`~.factor` does not
+reveal them. However, it may fail for some polynomials. Here are examples of
+{func}`~.roots`:
+
+```py
+>>> roots(expression, x)
+{-2: 2, 3: 1}
+>>> roots(symbolic, x)
+{-a: 2, b: 1}
+```
+
+It returns results as a dictionary, where the key is the root (for example, -2)
+and the value is the multiplicity of that root (for example, 2).
+
+{func}`~.roots` function uses a combination of techniques (factorization,
+decomposition, radical formulae) to find expressions in radicals if possible for
+the roots. When it can find some radical expressions for the roots, it returns
+them along with their multiplicity. This function will fail for most high-degree
+polynomials (five or greater) because they do not have radical solutions, and
+there is no guarantee that they have closed-form solutions at all, as explained
+by the [Abel-Ruffini
+theorem](https://en.wikipedia.org/wiki/Abel%E2%80%93Ruffini_theorem).
+
+#### {func}`~.real_roots`
+
+If the roots to your polynomial are real, using {func}`~.real_roots` ensures
+that only real (not complex) roots will be returned.
+
+```py
+>>> real_roots(expression)
+[-2, -2, 3]
```

Comment:
I was trying to add a similar example with the expression with symbolic constants, and get the error `PolynomialError: only univariate polynomials are allowed`:

```py
>>> from sympy import real_roots
>>> from sympy.abc import x, a, b
>>> symbolic = (x+a)**2 * (x-b)
>>> real_roots(symbolic)
PolynomialError                           Traceback (most recent call last)
/var/folders/9w/cn8tqbv976gckxd0c9cbr1sw0000gn/T/ipykernel_31641/1588471390.py in <module>
      2 a, b = symbols('a,b', real=True)
      3 # x = Symbol('x', positive=True)
----> 4 real_roots(symbolic)

~/opt/anaconda3/lib/python3.9/site-packages/sympy/polys/polytools.py in real_roots(f, multiple)
   6534             "can't compute real roots of %s, not a polynomial" % f)
   6535 
-> 6536     return F.real_roots(multiple=multiple)
   6537 
   6538 

~/opt/anaconda3/lib/python3.9/site-packages/sympy/polys/polytools.py in real_roots(f, multiple, radicals)
   3590 
   3591         """
-> 3592         reals = sympy.polys.rootoftools.CRootOf.real_roots(f, radicals=radicals)
   3593 
   3594         if multiple:

~/opt/anaconda3/lib/python3.9/site-packages/sympy/polys/rootoftools.py in real_roots(cls, poly, radicals)
    403     def real_roots(cls, poly, radicals=True):
    404         """Get real roots of a polynomial. """
--> 405         return cls._get_roots("_real_roots", poly, radicals)
    406 
    407     @classmethod

~/opt/anaconda3/lib/python3.9/site-packages/sympy/polys/rootoftools.py in _get_roots(cls, method, poly, radicals)
    736         """Return postprocessed roots of specified kind. """
    737         if not poly.is_univariate:
--> 738             raise PolynomialError("only univariate polynomials are allowed")
    739         # get rid of gen and it's free symbol
    740         d = Dummy()

PolynomialError: only univariate polynomials are allowed
```

I assume the issue is that `a` and `b` are being interpreted as variables. Is there a way to designate them as constants (not generators, I believe is the terminology)?

Reply:
The `real_roots` function uses `RootOf` so it can only work with univariate polynomials with rational coefficients. It can compute the roots of such a polynomial of any degree exactly though (unlike `roots`). In practice what this means is that you should substitute numbers for any symbolic parameters before calling `real_roots`.